testkit/jvm/src/main/scala/zio/test/Predicate.scala

```diff
@@ -0,0 +1,308 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.Exit
+
+/**
+ * A `Predicate[A]` is capable of producing assertion results on an `A`. As a
+ * proposition, predicates compose using logical conjuction and disjunction,
+ * and can be negated.
+ */
+class Predicate[-A] private (render: String, val run: A => PredicateResult) extends (A => PredicateResult) { self =>
+  import AssertResult._
+
+  /**
+   * Returns a new predicate that succeeds only if both predicates succeed.
+   */
+  final def &&[A1 <: A](that: => Predicate[A1]): Predicate[A1] =
+    Predicate.predicateDirect(s"(${self} && ${that})") { actual =>
+      self.run(actual) match {
+        case Failure(l) => Failure(l)
```

Comment:
Would this not unapply and recreate the object?
Would `f: Failure => f` be more efficient?

Reply:
Very true, but doesn't matter, the overhead of the test framework is inconsequential.